Mage/src/main/java/mage/abilities/effects/common/BalanceEffect.java

```diff
@@ -144,25 +129,17 @@ public boolean apply(Game game, Ability source) {
             if (player == null) {
                 continue;
             }
-            Cards cards = new CardsImpl();
-            TargetCardInHand target = new TargetCardInHand(minCard, filter3);
-            if (!target.choose(Outcome.Discard, player.getId(), source.getSourceId(), game)) {
+            TargetControlledPermanent target = new TargetControlledPermanent(min, min, filterPermanentDialog, true);
```

Comment:
Wrong logic. `Min` is card's amount to keep, not sacrifice. So you must use like `sacCount = countAll() - min`. Choose to sac.

BUT card's oracle text talk about choose to keep. So I recommends to change choose logic to it -- players must choose cards to keep (with good outcome for AI), not to sacrifice.